### PR TITLE
JVM_IR: box parameter type if needed

### PR DESCRIPTION
**Pull request by scaventz**
issue: https://youtrack.jetbrains.com/issue/KT-50986

### CODE REVIEW

File: compiler/testData/codegen/box/functions/kt50986.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// TARGET_BACKEND: JVM_IR
```
Comment (dnpetrov) [inline compiler/testData/codegen/box/functions/kt50986.kt:1]: Please, add `// CHECK_BYTECODE_LISTING` directive and re-run the test.
It'll produce `kt50986.txt` file containing generated bytecode signatures.
Check that it passes in the old back-end as well (to make sure that we don't break ABI).

**Comment by dnpetrov** [comment]
@ilmirus please, take a look at this issue in the context of value classes ABI.

**Comment by ilmirus** [comment]
Looks like the issue is fixed on the current master (714c5091cc2a19b7de4bb8ce586abc12b34e9d33) I cannot reproduce the issue.

**Comment by scaventz** [comment]
Indeed, the issue has been fixed by a7e2f7a6b6e906995f6d2f8c582406070873fc83 , I'm going to close this PR.

@dnpetrov Hi, it turns out I never understood the concept of `ABI` correctly, especially in the context of the Kotlin compiler, I googled it earlier but couldn't find an explanation related to the JVM platform. It's a little embarrassing to ask, but can you give me an example of how this PR possibly breaks the ABI and what are the potential risks of the breaking?

**Comment by dnpetrov** [comment]
@scaventz In context of this PR, I wanted to make sure that functions like
```
fun <T : Int> test1(x: T) {}
fun <T : Int> test2(x: T?) {}
fun <T : Int?> test3(x: T) {}
fun <T : Int?> test4(x: T?) {}
```
compile to methods with the same signatures.
Otherwise some code that invokes such function might stop working or fail to compile.
Life is life, such breaking changes are still possible if there's no other way around. We just have to go though our usual process for breaking changes, with language commitee making final decision regarding whether such change is acceptable, and what extra steps should be taken to mitigate such change.

**Pull request closed without merging**